### PR TITLE
Fix issue 3331 zero length reductions

### PR DESCRIPTION
**Pull request by crtrott**
This is the start of fixing zero length reductions corner cases described in #3331 . The first two commits add a test for RangePolicy zero length reductions into scalars, host and device views (for all backends) and the fix for the RangePolicy for CUDA. 

TODO: 
- Fix ROCM RangePolicy
- Add Test for MDRangePolicy and TeamPolicy
- Fix CUDA and ROCM MDRangePolicy and TeamPolicy
- Fix OpenMPTarget backend maybe?

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
Preliminary comment

**Comment by masterleinad** [comment]
There are still some unused variables that make CI checks fail:
```
/var/jenkins/workspace/Kokkos/core/unit_test/TestReduceCombinatorical.hpp:484:34: error: unused parameter 'N' [clang-diagnostic-unused-parameter]
  static void AddLambdaRange(int N, Kokkos::InvalidType, Args... /*args*/) {}
                                 ^
/var/jenkins/workspace/Kokkos/core/unit_test/TestReduceCombinatorical.hpp:487:33: error: unused parameter 'N' [clang-diagnostic-unused-parameter]
  static void AddLambdaTeam(int N, Kokkos::InvalidType, Args... /*args*/) {}
```

**Comment by crtrott** [comment]
Woohoo it passed!!

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 1]
```diff
@@ -1559,10 +1582,14 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
     }
 
     // Reduce with final value at blockDim.y - 1 location.
-    if (cuda_single_inter_block_reduce_scan<false, FunctorType, WorkTag>(
-            ReducerConditional::select(m_functor, m_reducer), blockIdx.x,
-            gridDim.x, kokkos_impl_cuda_shared_memory<size_type>(),
-            m_scratch_space, m_scratch_flags)) {
+    bool do_final_reduce = m_league_size == 0;
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1585]:
```suggestion
    bool do_final_reduce = (m_league_size == 0);
```

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 2]
```diff
@@ -1655,20 +1692,22 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
       dim3 grid(block_count, 1, 1);
       const int shmem_size_total = m_team_begin + m_shmem_begin + m_shmem_size;
 
+      if ((nwork < 1)
```
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1695]:
```suggestion
      if ((nwork == 0)
```
to look the same as above.

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 3]
```diff
@@ -742,11 +743,15 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
     }
 
     // Reduce with final value at blockDim.y - 1 location.
-    if (hip_single_inter_block_reduce_scan<false, FunctorType, work_tag>(
-            reducer_conditional::select(m_functor, m_reducer), blockIdx.x,
-            gridDim.x,
-            Kokkos::Experimental::kokkos_impl_hip_shared_memory<size_type>(),
-            m_scratch_space, m_scratch_flags)) {
+    bool do_final_reduce = m_league_size == 0;
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:746]:
```suggestion
    bool do_final_reduce = (m_league_size == 0);
```

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 4]
```diff
@@ -995,12 +1025,16 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
-    if (arg_policy.league_size() == 0) {
-      value_init::init(reducer_conditional::select(m_functor, m_reducer),
-                       m_result_ptr);
-      return;
-    }
+    // if (arg_policy.league_size() == 0) {
+    //   value_init::init(reducer_conditional::select(m_functor, m_reducer),
+    //                   m_result_ptr);
+    //  return;
+    //}
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:1037]: Why are you less certain here than for CUDA? I am not aware that we do anything much differently here. Anyway, this comment should go away one way or the other.

File: core/unit_test/TestReduceCombinatorical.hpp [hunk 5]
```diff
@@ -394,58 +394,69 @@ struct TestReduceCombinatoricalInstantiation {
...
 
-    value = 0;
+    value = 99;
 // WORKAROUND OPENMPTARGET Custom Reducers not implemented
 #ifndef KOKKOS_ENABLE_OPENMPTARGET
     CallParallelReduce(args...,
                        Test::ReduceCombinatorical::AddPlus<double>(value));
     if ((Kokkos::DefaultExecutionSpace::concurrency() > 1) &&
-        (ExecSpace::concurrency() > 1)) {
+        (ExecSpace::concurrency() > 1) && (expected_result > 0)) {
```
Comment (masterleinad) [inline core/unit_test/TestReduceCombinatorical.hpp:441]: Why do we need to guard the `N=0` case here?

File: core/unit_test/TestReduceCombinatorical.hpp [hunk 6]
```diff
@@ -553,95 +567,103 @@ struct TestReduceCombinatoricalInstantiation {
...
   template <class... Args>
-  static void AddPolicy_1(Args... args) {
-    int N = 1000;
+  static void AddPolicy_1(int N, Args... args) {
     Kokkos::RangePolicy<ExecSpace> policy(0, N);
 
-    AddFunctorLambdaRange(args..., 1000);
-    AddFunctorLambdaRange(args..., N);
-    AddFunctorLambdaRange(args..., policy);
+    AddFunctorLambdaRange(1000, args..., 1000);
```
Comment (masterleinad) [inline core/unit_test/TestReduceCombinatorical.hpp:573]: Should this run repeatedly independently of `N`?
```suggestion
```

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 7]
```diff
@@ -896,12 +918,16 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
-    if (m_league_size * m_team_size == 0) {
-      value_init::init(reducer_conditional::select(m_functor, m_reducer),
-                       arg_result.data());
-      return;
-    }
+    // if (m_league_size * m_team_size == 0) {
+    //  value_init::init(reducer_conditional::select(m_functor, m_reducer),
+    //                   arg_result.data());
+    //  return;
+    //}
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:930]: Why are you less certain here than for CUDA? I am not aware that we do anything much differently here. Anyway, this comment should go away one way or the other.

**Review by masterleinad — COMMENTED** [review]
Apart from some minor comments below this looks good to me.

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 7 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:930]: This was a reminder for you guys to tell me whether this is all kosher. If you say so we can delete these comments.

File: core/unit_test/TestReduceCombinatorical.hpp (hunk 6 above)
Comment (crtrott) [reply core/unit_test/TestReduceCombinatorical.hpp:573]: yeah it doesn't matter much. Otherwise I need to extract that in yet another different function.

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 4 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:1037]: As a reminder to you to check whether there is something which could go wrong ;-)

File: core/unit_test/TestReduceCombinatorical.hpp (hunk 5 above)
Comment (crtrott) [reply core/unit_test/TestReduceCombinatorical.hpp:441]: because for N==0 it would result in value==expected_result, while for every other case with N>0 (due to the way the team functor works) value>expected_result.

File: core/unit_test/TestReduceCombinatorical.hpp (hunk 6 above)
Comment (crtrott) [reply core/unit_test/TestReduceCombinatorical.hpp:573]: I.e. I need to add a AddPolicy_1a or so, i think running repeatedly that thing is fine.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 8]
```diff
@@ -1049,20 +1055,22 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
...
-#endif
 
       CudaParallelLaunch<ParallelReduce, LaunchBounds>(
           *this, grid, block, shmem,
           m_policy.space().impl_internal_space_instance(),
           false);  // copy to device and execute
 
       if (!m_result_ptr_device_accessible) {
-        Cuda().fence();
+        m_policy.space().fence();
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:1080]: See #3388.

File: core/unit_test/TestReduceCombinatorical.hpp (hunk 5 above)
Comment (masterleinad) [reply core/unit_test/TestReduceCombinatorical.hpp:441]: Yeah, it seems I forgot to delete this comment.

File: core/unit_test/TestReduceCombinatorical.hpp (hunk 6 above)
Comment (masterleinad) [reply core/unit_test/TestReduceCombinatorical.hpp:573]: Hmm... but we call AddPolicy_1 with N=1000 below so this check is the same as the next line, isn't it?

Comment (crtrott) [reply core/unit_test/TestReduceCombinatorical.hpp:573]: this is very explicitly checking literals vs runtime parameters.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: done

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 2 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp]: done

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 3 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp]: done

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 4 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp]: removed

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 7 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp]: removed

**Review by dalg24 — APPROVED** [review]
